### PR TITLE
Feature/block59 Library Package pipeline for devon4node

### LINKED ISSUE #59: Block: modify "Setup Library Package pipeline" for devon4node
**User Story:**
As a Developer (setting up the library package pipeline), I want to have a detailed guide and scripts for setting up Node.js environment, so that I can get it done fast and without errors.

**Details:**
Modify existing blocks related to pipelines with node.js related information. This includes updating guide and scripts related to "Setup Library Package pipeline​"

**Acceptance Criteria:**

1. Scripts are updated with Node.js related details
2. Guide is updated with Node.js related details

### PR DESCRIPTION
**Pull request by shiftwolf**
Closes #59

### CODE REVIEW

File: scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg [hunk 1]
```diff
@@ -12,33 +12,70 @@ sourceBranch="feature/library-package-pipeline"
...
+
+        # Create .npmrc
+        envsubst < .npmrc.template > "${localDirectory}/${scriptFilePath}/.npmrc"
+    ;;
+
+    "Maven")
+        export patToken
+
+        # Create settings.xml
+        envsubst < settings-template.xml > "${localDirectory}/${scriptFilePath}/settings.xml"
```
[the remark this answers is not in the document]
Comment (shiftwolf) [reply scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg:74]: This was pre-existing but I'll change it accordingly.

File: documentation/azure-devops/setup-library-package-pipeline.asciidoc [hunk 2]
```diff
@@ -54,24 +54,35 @@ NOTE: The config file for the library package pipeline is located at `/scripts/p
...
 
+=== npm feed
+Add a .npmrc to your project, in the same directory as your package.json:
+
+```
+registry=https://pkgs.dev.azure.com/<yourOrganization>/_packaging/<yourFeed>/npm/registry/
+
+always-auth=true
+```
+
```
[the remark this answers is not in the document]
Comment (shiftwolf) [reply documentation/azure-devops/setup-library-package-pipeline.asciidoc:77]: Ok, in retrospective this seems very obvious. If you have any suggestions for naming conventions that would be helpful, as I would need 2 different ".npmrc.template" files. In the meantime I'll go with ".npmrc.project.template" for this one and ".npmrc.user.template" for the user-level npmrc.

[the remark this answers is not in the document]
Comment (mcerverc) [reply documentation/azure-devops/setup-library-package-pipeline.asciidoc]: The naming convention you chose seems proper for me :)

**Review by mcerverc — APPROVED** [review]

File: scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg [hunk 3]
```diff
@@ -12,33 +12,73 @@ sourceBranch="feature/library-package-pipeline"
...
+        export email
+
+        encodedPatToken=$(python -c "import base64; print(base64.b64encode('${patToken}'.encode('ascii')).decode('ascii'))")
+        export encodedPatToken
+
+        # Create user-level .npmrc
+        envsubst < .npmrc.user.template > "${localDirectory}/${scriptFilePath}/.npmrc"
+
+        # Create project .npmrc
+        envsubst < .npmrc.project.template > "${localDirectory}/.npmrc"
```
Comment (ultymatom) [inline scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg:70]: Be careful this file is not being commited

File: scripts/pipelines/azure-devops/templates/library-package/.npmrc.user.template [hunk 4]
```diff
@@ -0,0 +1,8 @@
+; begin auth token
+//pkgs.dev.azure.com/${organization}/_packaging/${organization}/npm/registry/:username=$organization
+//pkgs.dev.azure.com/${organization}/_packaging/${organization}/npm/registry/:_password=$encodedPatToken
+//pkgs.dev.azure.com/${organization}/_packaging/${organization}/npm/registry/:email=$email
+//pkgs.dev.azure.com/${organization}/_packaging/${organization}/npm/:username=$organization
+//pkgs.dev.azure.com/${organization}/_packaging/${organization}/npm/:_password=$encodedPatToken
+//pkgs.dev.azure.com/${organization}/_packaging/${organization}/npm/:email=$email
+; end auth token
```
Comment (ultymatom) [inline scripts/pipelines/azure-devops/templates/library-package/.npmrc.user.template:8]: $email and $encodedPatToken are not being replaced when I execute it, when seeing your code I don't see anything wrong because you export corectly the values before the envsubst, was this working when you fully tested the script ?

Comment (shiftwolf) [reply scripts/pipelines/azure-devops/templates/library-package/.npmrc.user.template:8]: I'll look into it.

File: scripts/pipelines/azure-devops/templates/library-package/.npmrc.user.template [hunk 5]
```diff
@@ -0,0 +1,8 @@
+; begin auth token
+//pkgs.dev.azure.com/${organization}/_packaging/${organization}/npm/registry/:username=$organization
+//pkgs.dev.azure.com/${organization}/_packaging/${organization}/npm/registry/:_password=$encodedPatToken
+//pkgs.dev.azure.com/${organization}/_packaging/${organization}/npm/registry/:email=$email
+//pkgs.dev.azure.com/${organization}/_packaging/${organization}/npm/:username=$organization
+//pkgs.dev.azure.com/${organization}/_packaging/${organization}/npm/:_password=$encodedPatToken
+//pkgs.dev.azure.com/${organization}/_packaging/${organization}/npm/:email=$email
```
Comment (ultymatom) [inline scripts/pipelines/azure-devops/templates/library-package/.npmrc.user.template:7]:
```suggestion
//pkgs.dev.azure.com/${organization}/_packaging/${feed}/npm/registry/:username=$organization
//pkgs.dev.azure.com/${organization}/_packaging/${feed}/npm/registry/:_password=$encodedPatToken
//pkgs.dev.azure.com/${organization}/_packaging/${feed}/npm/registry/:email=$email
//pkgs.dev.azure.com/${organization}/_packaging/${feed}/npm/:username=$organization
//pkgs.dev.azure.com/${organization}/_packaging/${feed}/npm/:_password=$encodedPatToken
//pkgs.dev.azure.com/${organization}/_packaging/${feed}/npm/:email=$email
```

The URL to access the feed was not correct

Comment (shiftwolf) [reply scripts/pipelines/azure-devops/templates/library-package/.npmrc.user.template:7]: Just to confirm, this fixes the issue with the $email and $patToken not being replaced?

File: documentation/azure-devops/setup-library-package-pipeline.asciidoc [hunk 6]
```diff
@@ -54,6 +54,11 @@ NOTE: The config file for the library package pipeline is located at `/scripts/p
 ./pipeline_generator.sh -c ./templates/library-package/library-package-pipeline.cfg -n library-package-pipeline -l java -d C:/Users/$USERNAME/Desktop/java-library-project --build-pipeline-name java-library-build --quality-pipeline-name java-library-quality
 ```
 
+==== Node.js project
+```
+./pipeline_generator.sh -c ./templates/library-package/library-package-pipeline.cfg -n library-package-pipeline -l node -d C:/Users/$USERNAME/Desktop/node-library-project --build-pipeline-name node-library-build --quality-pipeline-name node-library-quality
```
Comment (ultymatom) [inline documentation/azure-devops/setup-library-package-pipeline.asciidoc:59]: in the pipeline-generator.sh some of the argument are not written in the help function (I think it is an error that comes from before, but better adding them)

File: scripts/pipelines/azure-devops/templates/library-package/.npmrc.user.template (hunk 4 above)
Comment (ultymatom) [reply scripts/pipelines/azure-devops/templates/library-package/.npmrc.user.template:8]: forget about it, now without any changed from it is working correctly^^

File: scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg [hunk 7]
```diff
@@ -12,33 +12,73 @@ sourceBranch="feature/library-package-pipeline"
...
-    #Create PAT
+    # Create PAT
     echo -e "${green}Creating a PAT with necessary permissions..."
     patToken=$(az rest --method post --uri "https://vssps.dev.azure.com/${organization}/_apis/tokens/pats?api-version=7.1-preview.1" --resource "https://management.core.windows.net/" --body '{ "displayName": "libraryDeployToken", "validTo": "2099-02-25T11:44:36.1966667Z", "scope": "vso.packaging_manage" }')
     patToken=$(echo "$patToken" | python -c  "import sys, json; print(json.load(sys.stdin)['patToken']['token'])")
 
-    #Create feed
-    echo -e "${green}Creating a Maven feed..."
+    # Create feed
+    echo -e "${green}Creating a ${feedType} feed..."
```
Comment (ultymatom) [inline scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg:49]:
```suggestion
    echo -e "${green}Creating a ${feedType} feed...${white}"
```

Without changing back to white all the next lines are green

File: documentation/azure-devops/setup-library-package-pipeline.asciidoc (hunk 6 above)
Comment (ultymatom) [inline documentation/azure-devops/setup-library-package-pipeline.asciidoc:59]: If I am not wrong I am not sure you need to have the build pipeline as an argument. Because we use the --build-pipeline-name in case of a java library so that we can download the application that have been built before, in the node case I am not sure this is needed. (in that case the yml.template needs to be updated with conditional step)

File: scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg [hunk 8]
```diff
@@ -12,33 +12,73 @@ sourceBranch="feature/library-package-pipeline"
 pipelinePath=".pipelines"
 # Path to the scripts.
```
Comment (ultymatom) [inline scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg:13]: (I put a comment here but this is related to the line with mandatory flags but github does not let me comment on that line)

if as said on another comment we don't need the --build-pipeline-name flag, you will probably need to set the variable mandatoryFlags conditionally

File: scripts/pipelines/azure-devops/templates/library-package/.npmrc.user.template (hunk 5 above)
Comment (ultymatom) [reply scripts/pipelines/azure-devops/templates/library-package/.npmrc.user.template:7]: Actually after I did these changes in local it was working but I don't know how ^^
but I put the comment about the two variables not being replaced a resolved

Comment (shiftwolf) [reply scripts/pipelines/azure-devops/templates/library-package/.npmrc.user.template:7]: Also, I'm not sure how this could happen. At some point I was able to make it work (or it appeared so at least) but I don't know what went wrong from there. I'll definitely do much more (regression) testing in the future. Thanks for the help in finding this issue though.

File: scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg (hunk 3 above)
Comment (shiftwolf) [reply scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg:70]: As far as I have tested now this is not a problem, at least from the npm side of things, git may be different. But to be absolutely sure I will try to add a line to both the .npmignore and .git/info/exclude files to explicitly ignore .npmrc files.

Comment (shiftwolf) [reply scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg:70]: As far as I have tested now this is not a problem, at least from the npm side of things, git may be different. But to be absolutely sure I will try to add a line to both the .npmignore and .git/info/exclude files to explicitly ignore .npmrc files.

Comment (ultymatom) [reply scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg:70]: it will work in local but your pipeline won't be working because you won't have teh .npmrc at the root of your repo so the URL that npm publish will use is not the righ one

**Comment by shiftwolf** [comment]
I have tested also on Azure DevOps but I will continue to test this because there may be some edge cases that are not working.

The main problem currently is that once this uploaded to Azure DevOps, it will break both the build and the quality pipeline, as they can't handle the `.npmrc` that is generated here. So far the only solutions I was able to find were changes to the build and quality pipelines themselves, like adding an additional parameter to use the npm public registry:  `--registry=https://registry.npmjs.org/` which causes the `.npmrc` contents to be ignored. Would that be a viable solution?

File: scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg [hunk 9]
```diff
@@ -1,5 +1,3 @@
-# Mandatory flags.
-mandatoryFlags="$pipelineName,$localDirectory,$language,$buildPipelineName,$qualityPipelineName,"
```
Comment (ultymatom) [inline scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg:2]: Be careful, here you you removed the mandatory flag variable,
you still need tha var, I meant define it conditionally as in the package pipeline:
https://github.com/devonfw/hangar/blob/master/scripts/pipelines/azure-devops/templates/package/package-pipeline.cfg

In that file we defined conditionally the mandatory flags depending on if we use AWS to store the image or another registry.

(don't hesitate to contact me on teams if anything in unclear , (I'm Timothé Paty for info))

Comment (ultymatom) [reply scripts/pipelines/azure-devops/templates/library-package/library-package-pipeline.cfg:2]: Indeed as you said me, the flags are defined lower in the file so my commentary makes no sense

File: scripts/pipelines/azure-devops/pipeline_generator.sh [hunk 10]
```diff
@@ -83,6 +83,8 @@ function help {
     echo ""
     echo "Library package pipeline flags:"
     echo "  -l, --language              [Required] Language or framework of the project."
+    echo "      --quality-pipeline-name [Required] Quality pipeline name."
+    echo "      --build-pipeline-name   [Required, if Java / Maven] Build pipeline name."
```
Comment (ultymatom) [inline scripts/pipelines/azure-devops/pipeline_generator.sh:87]: I'm writting here because I can't put it on the line of the docs, but you have to write in the docs that it is not mandatory the build pipeline name (in the two lines it gives info about it)

Comment (shiftwolf) [reply scripts/pipelines/azure-devops/pipeline_generator.sh]: I hope my latest change resolved this.

**Review by ultymatom — APPROVED** [review]

File: documentation/azure-devops/setup-library-package-pipeline.asciidoc [hunk 11]
```diff
@@ -3,15 +3,15 @@ toc::[]
 :idprefix:
 :idseparator: -
 
-= Setting up a Java library package pipeline on Azure DevOps
+= Setting up a library package pipeline using Azure DevOps
```
Comment (albertdb) [inline documentation/azure-devops/setup-library-package-pipeline.asciidoc:6]:
```suggestion
= Setting up a library package pipeline on Azure DevOps
```

File: documentation/azure-devops/setup-library-package-pipeline.asciidoc [hunk 12]
```diff
@@ -3,15 +3,15 @@ toc::[]
...
 
-In this section we will create a package pipeline to generate a distributable of a Java library and push it to a Maven feed in Azure DevOps. This pipeline will be configured in order to be triggered every time quality pipeline is executed successfully on a commit for `release/*` and `develop` branches, requiring manual launch for other branches but still enforcing that quality pipeline has passed. It consumes the artifact produced by the build pipeline.
+In this section we will create a package pipeline to generate a distributable of a library and push it to a feed in Azure DevOps. This pipeline will be configured in order to be triggered every time quality pipeline is executed successfully on a commit for `release/*` and `develop` branches, requiring manual launch for other branches but still enforcing that quality pipeline has passed. When Java is used, it also consumes the artifact produced by the build pipeline.
```
Comment (albertdb) [inline documentation/azure-devops/setup-library-package-pipeline.asciidoc:8]:
```suggestion
In this section we will create a package pipeline to generate a distributable of a library and push it to a feed in Azure DevOps. This pipeline will be configured in order to be triggered every time quality pipeline is executed successfully on a commit for `release/*` and `develop` branches, requiring manual launch for other branches but still enforcing that quality pipeline has passed. Depending on the language, it consumes the artifact produced by the build pipeline.
```

**Review by albertdb — CHANGES_REQUESTED** [review]

**Review by albertdb — APPROVED** [review]
